HTML API: Handle parsing changes in foreign content.

dmsnell opened the pull request:
Trac ticket: Core-61576

## Status

This is undergoing final review and preparation.

### Earlier versions
 - First draft (ba2c8896f08ee36d3c61a500a86422b736a00923)
   - By implementing the rules at the top of Tree Construction/Dispatch, we don't need `$skip_next_foreign_content_processing`, because we pop nodes off of the stack of open elements until we're back at an `html` element, or an HTML integration point, or a MathML integration point. Returning to `step()` will choose _insertion mode_ instead of _foreign content_ and prevent an infinite loop.

 - Second draft (3e4bb4d011100c3a5641ebffd1beb09b0d808f15)
   - In this draft the attribute and tag names are properly cased, but it throws off the logic mechanisms. Since attributes and tags are still parsed according to HTML rules (ASCII-case-insensitive), it makes sense to continue reporting them in all caps. Perhaps a new method such as `get_token_display_name()` would provide the remapping.

 - Third draft (@sirreal's work)
   - Jon found a way to fix the namespace issues by rearranging the assignment of a namespace to the tokens and the detection of integration points.

### Notes
 - There's an interplay we have to figure out with casing of tag names. For all of our comparison functions we check the upper-case variants, but there's value in reporting the foreign content tag names as lower-cased or as mixed case.
 - There are some SVG tag names that are supposed to report in mixed case. For both SVG and MathML there are attributes which are supposed to report in mixed case. Unlike my first assumption, however, they still all parse with ASCII case insensitivity. That is, `<circle clippath=1 clipPath=2 CLIPPATH=3 cliPPath=4>` has a single attribute named `clipPath` and whose value is `1`. This is convenient because we don't have to change the Tag Processor, but inconvenient when things are mostly based on lower-case attribute names.
 - We probably need to repeal the idea that tag names are upper-case and add another bit to communicate tag name vs. doctype declaration. Maybe? Can we test that `math:mi` or `math:MI` is in the `math` namespace? There can be no HTML tag named `math:mi` (it would be `MATH:MI`).
 - Could be handy to have something like `base_class_get_tag()` or `private function comparison_tag_name()` etc… to report an upper-case tag name, while preserving the case-variants required in foreign content to outside calls.
 - Tracking the namespace seems a bit fishy. When we're in the Tag Processor we can know that if we encounter an SVG or MATH tag when we're in the `html` namespace, that it should change the namespace to `svg` or `math`, respectively, and lower-case the tag names. However, the role of integration points and parsing things in the insertion mode is still vague.
 - Foreign content _is not_ an insertion mode. This is important when handling HTML tags inside HTML integration points. This is _**very important**_ for `get_modifiable_text()` which doesn't know if a text node inside foreign content is being processed _as foreign content_ or _in the insertion mode_, where it determines if NULL bytes should be replaced or removed.
 - It's probably necessary to start over again with these insights and more in mind.

## Description

We should reliable detect foreign content and we need to do it in the Tag Processor, specifically because of the rules for CDATA sections. The HTML Processor needs this as well to determine if things like self-closing flags for HTML elements should be respected.

```diff
- Tests: 1498, Assertions: 787, Skipped: 711.
+ Tests: 1498, Assertions: 930, Skipped: 568.
```

**Unlocks**:
 - proper detection of CDATA
 - parsing of SVG and MathML
 - indicating if one should expect a closing tag or not

<img width="918" alt="Screenshot 2024-02-01 at 9 49 43 PM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/00ca7c32-d269-4da7-a9ce-04aff9640f8a">

![Screenshot 2024-02-02 at 12 13 47 AM](https://github.com/WordPress/wordpress-develop/assets/5431237/7a104953-9dbb-48e4-ae45-be40f3df9736)

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -1725,6 +1739,32 @@ private function parse_next_tag() {
+					'A' === $html[ $at + 7 ] &&
+					'[' === $html[ $at + 8 ]
+				) {
+					$closer_at = strpos( $html, ']]>', $at + 1 );
```
Minor, but this could start checking after we find the cdata opener:

```suggestion
					$closer_at = strpos( $html, ']]>', $at + 9 );
```

dmsnell committed:
HTML API: Add support for SVG and MathML (Foreign content)

As part of work to add more spec support to the HTML API, this patch adds
support for SVG and MathML elements, or more generally, "foreign content."

The rules in foreign content are a mix of XML and HTML parsing rules and
introduce additional complexity into the processor, but is important in
order to avoid getting lost when inside these elements.

Developed in https://github.com/wordpress/wordpress-develop/pull/6006
Discussed in https://core.trac.wordpress.org/ticket/61576

Props: dmsnell, jonsurrell.
See #61576.

dmsnell committed:
Update test runner

dmsnell committed:
Lower-case tag names for foreign content in test runner.

dmsnell committed:
Use the namespace-adjusted tag name in tests.

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -777,7 +773,9 @@ public function expects_closer( $node = null ): ?bool {
-			in_array( $token_name, array( 'IFRAME', 'NOEMBED', 'NOFRAMES', 'SCRIPT', 'STYLE', 'TEXTAREA', 'TITLE', 'XMP' ), true )
+			in_array( $token_name, array( 'IFRAME', 'NOEMBED', 'NOFRAMES', 'SCRIPT', 'STYLE', 'TEXTAREA', 'TITLE', 'XMP' ), true ) ||
+			// Self-closing elements in foreign content.
+			( isset( $node ) && 'html' !== $node->namespace && $node->has_self_closing_flag )
```
Unless there is a performance benefit for doing this, I think generally better to check for equality with `null` since the variable _is_ set. But even better here to check if it is the instance of the expected class, right?

```suggestion
			( $node instanceof WP_HTML_Token && 'html' !== $node->namespace && $node->has_self_closing_flag )
```

dmsnell replied:
yeah. I think I did this for a reason, but I can't remember. likely I had other code in transition and had non-null non-token instances passed around

sirreal replied:
Bookmarking this, the `isset( $node )` seems like it would make this function return `true` under some circumstances where we would not expect that.

dmsnell replied:
there was much discussion over this little function, and a lot of effort to "fix" it that only resulted in broken code.

I'd like to propose we continue with it as written and come back afterwards to clean it up.

right now, all of the "fixes" are only used in test code when `expects_closer()` is called without a `$node` passed in. the real problem is that there's confusion in what this function is communicating and how `step()` pops elements off of the stack of open elements.

we recently realized that we need to pop elements off not in `step()`, but in the parsing rules where they indicate as much. I'd rather we ship this odd bit now and create the more-sweeping fix in one go instead of stewing on code we think must be wrong but which we don't have a failing test case for.

---

I've added a type annotation to the function, ensuring that `isset()` is sufficient, and I've added a `@todo` annotation earmarking this for review. I _think_ what's going on is a complicated interaction with `step()` where this looks wrong but isn't.

By the way, this function returns "expects closer" for all closing tags, which is "wrong" but also part of the bigger issue not created in this PR.

Thoughts?

sirreal replied:
PR to address this: https://github.com/WordPress/wordpress-develop/pull/7162

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2237,8 +2260,12 @@ private function step_in_body(): bool {
 
-				$this->bail( 'Cannot process MATH element, opening foreign content.' );
-				break;
+				$this->change_parsing_namespace( 'math' );
```
Where is the parsing namespace changed back to `html` when encountering `</SVG>` or `</MATH>`?

dmsnell replied:
I'm having trouble figuring out from the spec where the namespace is assigned for any other token besides `SVG` and `MATH`, as I'm having failures with a `math ANNOTATION-XML` tag

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -3419,7 +3444,205 @@ private function step_after_after_frameset(): bool {
+		 * > An end tag whose name is "script", if the current node is an SVG script element.
+		 */
+		if ( $this->is_tag_closer() && 'SCRIPT' === $this->state->current_token->node_name && 'svg' === $this->state->current_token->namespace ) {
+			$this->bail( 'Cannot parse SCRIPT tags inside SVG elements.' );
```
Can't this just be ignored since the scripting flag is not enabled?

dmsnell replied:
it can't be ignored because there are rules that we would (might) need to follow and I didn't want to dive into that _at this point_. perhaps the rules aren't that complicated. it would require looking into the spec, determining what is required, and if we have enough state tracking already to handle it.

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4483,18 +4761,106 @@ private function insert_virtual_node( $token_name, $bookmark_name = null ): void
+		if ( 'svg' === $this->get_namespace() ) {
+			return (
+				'DESC' === $tag_name ||
+				'FOREIGNoBJECT' === $tag_name ||
```
Why the lower-case `o`?

dmsnell replied:
just a TyPO 😄

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4483,18 +4761,106 @@ private function insert_virtual_node( $token_name, $bookmark_name = null ): void
 	 */
 	public static function is_special( $tag_name ): bool {
-		$tag_name = strtoupper( $tag_name );
+		if ( is_string( $tag_name ) ) {
+			$tag_name = strtoupper( $tag_name );
+		} else {
+			$tag_name = 'html' === $tag_name->namespace
+				? strtoupper( $tag_name->node_name )
+				: "{$tag_name->namespace} {$tag_name->node_name}";
+		}
```
Since the normal case is now the object:

```suggestion
		if ( $tag_name instanceof WP_HTML_Token ) {
			$tag_name = 'html' === $tag_name->namespace
				? strtoupper( $tag_name->node_name )
				: "{$tag_name->namespace} {$tag_name->node_name}";
		} else {
			$tag_name = strtoupper( $tag_name );
		}
```

Also, should `$tag_name` be renamed `$node`?

dmsnell replied:
yeah probably. thanks for the review - there are still some logical error in here I'm tracking, specifically, trying to figure out the right place to assign namespaces and avoid infinite loops when reprocessing nodes _from_ the foreign content rules.

westonruter reviewed (commented):
❤️

dmsnell committed:
Merge branch 'trunk' into html-api/detect-foreign-content

dmsnell committed:
Various updates and changes. Currently experiencing infinite loops.

dmsnell commented:
@sirreal @westonruter I've pushed some updates that incorporate `trunk` and a few other fixes. notably, I had the wrong code for `adjusted_current_node()`.

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -3862,7 +3882,210 @@ private function step_after_after_frameset(): bool {
+						break;
+					}
+
+					echo "\e[90mPopping a \e[35m{$current_node->namespace} \e[34m{$current_node->node_name}\e[90m from the stack.\e[m\n";
```
Debug code I'm presuming.

```suggestion
```

westonruter commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -3862,7 +3882,210 @@ private function step_after_after_frameset(): bool {
+
+			$adjusted_current_node = $this->get_adjusted_current_node();
+			$this->insert_foreign_element( $this->state->current_token, $adjusted_current_node->namespace, false );
+			echo "\e[90mACI is a \e[35m{$this->state->current_token->namespace}\e[90m \e[34m{$this->state->current_token->node_name}\e[m\n";
+			echo "\e[90mInserted a \e[35m{$this->state->current_token->namespace}\e[90m \e[34m{$this->state->current_token->node_name}\e[m\n";
```
Debug code I'm presuming.

```suggestion
```

dmsnell committed:
Prevent infinite loop.

dmsnell committed:
Check namespace against actual token, not Tag Processor state.

dmsnell committed:
Provide method for getting adjusted attribute names, and update testing code.

sirreal committed:
HTML5Lib: Prevent printing empty text nodes in tree

sirreal commented:
This is very tricky. I'll share some cases that are problematic right now:




**Update:** I've debugged all of these cases and pushed some changes to fix them.

<details>
<summary>Fixed</summary>


### `<svg><a/>Text after "svg:a" self closing tag.`
Errors with `Cannot run adoption agency when "any other end tag" is required.`

### `<svg></svg><![CDATA[ bogus comment ]]>`
![svg :svg ! CDATA  bogus comment ](https://github.com/user-attachments/assets/c426ad17-e2a1-4ce2-8bdf-46096c7399b7)

### `<svg><foreignObject><![CDATA[ bogus comment ]]>`
![svg foreignObject ! CDATA  bogus comment ](https://github.com/user-attachments/assets/f7a953f2-c71e-4cbf-bd3e-7abb630e9dac)

### `<svg><title><div>`
![svg title div](https://github.com/user-attachments/assets/932b45c8-b035-436f-b8af-23941e817481)

### `<svg><title><rect><div>`
![svg title rect div](https://github.com/user-attachments/assets/c1d74e64-5dda-4882-b435-aff720873175)

### `<svg><title><svg><div>`
![svg title svg div](https://github.com/user-attachments/assets/370a1b4a-4945-434d-b58c-22f7e88030be)



### `<math><mi><div></div></mi><mi>`
![math mi div :div :mi mi](https://github.com/user-attachments/assets/236169af-22d1-4671-a554-7401bc85f596)

### `<math><mi><svg><foreignObject><div>`
![math mi svg foreignObject div](https://github.com/user-attachments/assets/fe4ce497-2ce5-45e4-bbf8-7edf3a8873b9)

### `<svg><foreignObject>\x00filler\x00text` (null bytes)
![svg foreignObject x00filler x00text](https://github.com/user-attachments/assets/bc5a28eb-d407-418b-8be3-a321c1a4a292)


### `<svg></br><foo>`
![svg :br foo](https://github.com/user-attachments/assets/33c56197-f15b-45f7-b304-3449319f77df)

### `<math><mtext><b>`
![math mtext b](https://github.com/user-attachments/assets/ab38da64-cdc1-424c-b25a-c4d547bd5089)

</details>

sirreal committed:
phpcs fix

dmsnell committed:
Closing BR can occur outside of the HTML namespace.

sirreal committed:
Fix an issue with mathml integration point check

sirreal committed:
Change parsing namespace in push/pop handlers

The push/pop handlers are a good opportunity to change parsing namespace
as elements are encountered and moved onto the stack. This helps to keep
the parsing namespace up to date without requiring state changes spread
all over parsing code.

sirreal committed:
Check for namespace in expects_closer atomic elements

This fixes an issue where <title> tags in SVG should not immediately pop off the stack.

sirreal committed:
Use insert_foreign_element to insert foreign elements

sirreal commented:
I've made some progress on this fixing some of the issues mentioned above: https://github.com/WordPress/wordpress-develop/pull/6006#issuecomment-2262585496

Right now some self-closing tags in foreign content break. HTML like `<svg><a/>` seems to enter step_in_body, runs the adoption agency algorithm, and fails:

```
WP_HTML_Unsupported_Exception: Cannot run adoption agency when "any other end tag" is required. in /var/www/html/wp-includes/html-api/class-wp-html-processor.php:450
Stack trace:
#0 /var/www/html/wp-includes/html-api/class-wp-html-processor.php(5145): WP_HTML_Processor->bail('Cannot run adop...')
#1 /var/www/html/wp-includes/html-api/class-wp-html-processor.php(2423): WP_HTML_Processor->run_adoption_agency_algorithm()
#2 /var/www/html/wp-includes/html-api/class-wp-html-processor.php(4108): WP_HTML_Processor->step_in_body()
#3 /var/www/html/wp-includes/html-api/class-wp-html-processor.php(889): WP_HTML_Processor->step_in_foreign_content()
#4 /var/www/html/wp-includes/html-api/class-wp-html-processor.php(635): WP_HTML_Processor->step()```

sirreal committed:
Do not report foreign content self-closing tags as closers

These tags are openers _and_ closers, which confuses things. Treating
them explicitly as closers seems more confusing.

Maybe some multi-state return value would be best.

Void tags do not report themselves as closers, which is basically the
same.

sirreal committed:
Use get_namespace() over parsing_namespace

This fixes some issues where the parsing_namespace has changed
for continued processing, but the current token is in a
different namespace.

sirreal committed:
Use get_adjusted_current_node when getting namespace

This fixes some issues where the parsing namespace may have changed or
non-element nodes like #text do not have a meaningful namespace.

sirreal committed:
Fix adjusted_current_node to check _element_ nodes

The stack of open elements may have non-element nodes, This causes
problems with algorithms that check for the number of nodes on the stack
or check the namespace of the current node.

This adds some new methods to the stack of open elements, but a better
solution is to only include elements on the stack of open elements.

sirreal committed:
Fix typo in femorphology tag name

sirreal committed:
More passing FC tests

sirreal committed:
Only transform attributes outside of html namespace

Some attribute name transformations were applied in all namespaces, but
should only be applied in foreign content.

dmsnell commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -2732,19 +3188,25 @@ public function has_self_closing_flag(): bool {
+
+			case 'math':
+			case 'svg':
+				return $this->is_closing_tag;
```
@sirreal why remove `has_self_closing_flag()` from this? if it's present, I was considering that a closing tag.

maybe the terminology is skewed when it comes to self-closing elements. it _does_ close the tag, but it is simultaneously an opener. to me, `is_tag_closer()` captures that, but maybe that's confusing.

perhaps this is best handled in `expects_closer()`

sirreal replied:
It's unclear exactly how these should behave. This change fixed an issue where this condition was not being entered on self-closing tags:

https://github.com/WordPress/wordpress-develop/blob/f66ba57cd96c6d0e38e81d90bf91134423d7844e/src/wp-includes/html-api/class-wp-html-processor.php#L4018-L4042

If this condition is included, it becomes difficult to know if a tag is intended to create an element, so the condition above would have to become something like `! $this->is_tag_closer() && ! $this->has_self_closing_flag()` in order to know that it _is_ a tag opener.

---

I'm not convinced one way or another.

On one hand, it's a tag opener and a tag closer, so reporting that it's a tag closer is correct.
On the other hand, it feels like they're _essentially_ openers and not closers. They happen to be self-close, but that's not what we're really asking.

Maybe we need a new method, or maybe that one place I mentioned above just needs to be updated to account for self-closers being reported as closers.

dmsnell replied:
after talking about this I think you're right. this should, inevitably, be handled in `expects_closer()`, not in `is_tag_closer()`

sirreal committed:
Merge branch 'trunk' into html-api/detect-foreign-content

sirreal committed:
Fix svg:TEMPLATE handling in html5lib tests

Template tags in the SVG namespace _should not_ behave
as html template tags.

sirreal committed:
Fix HTML5lib-test handling of self-closing tags

Tags in MathML and svg namespaces should respect self-closing.
Fix the indentation adjustment on self-closing tags when constructing
the tree representation.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -4069,7 +4088,289 @@ private function step_after_after_frameset(): bool {
+			$this->insert_foreign_element( $this->state->current_token, $this->state->current_token->namespace, false );
+
+			if ( $this->state->current_token->has_self_closing_flag ) {
+				if ( 'SCRIPT' === $this->state->current_token->node_name && 'svg' === $this->state->current_token->namespace ) {
+					$this->bail( 'Cannot parse SCRIPT tags inside SVG elements.' );
```
Will you explain this? `<svg><script></script>` is handled, but `<svg><script/>` bails, was that intentional?

dmsnell replied:
Explained in fdf36bc

sirreal committed:
Do not need to worry about those elements after all

sirreal committed:
Move namespace handling into insert_foriegn_element

sirreal committed:
Handle namespace mucking inside "insert_foreign_element"

sirreal committed:
Revert "Handle namespace mucking inside "insert_foreign_element""

This reverts commit e0046034ab390d1e758b03bf0f9b1f710b8d5db5.

sirreal committed:
Revert "Move namespace handling into insert_foriegn_element"

This reverts commit 9a83dc7aca9c9c0c93c8c3d98687c1d1a6959b05.

sirreal committed:
Revert irrelevant change

sirreal committed:
Reapply "Move namespace handling into insert_foriegn_element"

This reverts commit c944b1b573856cac4be1b609906354d73561a50e.

sirreal committed:
Reapply "Handle namespace mucking inside "insert_foreign_element""

This reverts commit 641e678e35dd625f69b1977bb8ff95c5e4b3d23f.

sirreal committed:
Only insert foreign elements when handling foreign content

sirreal committed:
Do not change the namespace to the integration node type

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -2741,19 +3200,25 @@ public function has_self_closing_flag(): bool {
+				 *
+				 * @see https://html.spec.whatwg.org/#parsing-main-inbody
+				 */
+				return $this->is_closing_tag && 'BR' !== $this->get_tag();
```
This can probably be reverted because BR tags will close out of foreign content.

dmsnell replied:
Reverted in f4373ff

dmsnell committed:
Update comments, reorder `try`, support SCRIPT tags.

dmsnell committed:
Revert change to BR closing tag behavior.

dmsnell committed:
Rename "namespaced" tag/attribute name getters to "get_qualified_...()"

dmsnell committed:
WPCS

dmsnell committed:
New TODO ear-marking a logic review.

sirreal committed:
Move the foreign content other end tag label into condition

The specification says to run run the "any other end tag" steps under
certain other conditions. A goto and label are used for this. When
moving to the label, the "any other end tag" condition does not apply,
the goal is to run the steps regardless.

Move the label into the conditional block so that the condition is not
erroneously checked.

This caused a failure to return a value when stepping on
`<svg><script/>`.

sirreal reviewed (approved):
This looks ready to me, I'm very excited to add this support!

The amount of html5lib-tests that run (and all pass) gives me confident in this work:

```
OK, but incomplete, skipped, or risky tests!
Tests: 1498, Assertions: 930, Skipped: 568.
```

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/detect-foreign-content

dmsnell commented:
Merged in [[58867]](https://core.trac.wordpress.org/changeset/58867)
https://github.com/wordpress/wordpress-develop/commit/de084d7d0e302027fb8f0a99bbeedf88e079efee
